flink/v1.12/flink/src/test/java/com/netease/arctic/flink/read/hidden/kafka/TestKafkaSourceReader.java

```diff
@@ -0,0 +1,258 @@
+import com.netease.arctic.log.LogData;
+import com.netease.arctic.log.LogDataJsonDeserialization;
+import com.netease.arctic.log.LogDataJsonSerialization;
+import com.netease.arctic.utils.IdGenerator;
+import org.apache.flink.api.common.eventtime.Watermark;
+import org.apache.flink.api.connector.source.ReaderOutput;
+import org.apache.flink.api.connector.source.SourceOutput;
+import org.apache.flink.api.connector.source.SourceReader;
+import org.apache.flink.connector.kafka.source.enumerator.initializer.OffsetsInitializer;
+import org.apache.flink.connector.kafka.source.split.KafkaPartitionSplit;
+import org.apache.flink.connector.testutils.source.reader.TestingReaderContext;
+import org.apache.flink.table.data.RowData;
+import org.apache.kafka.clients.consumer.ConsumerRecords;
+import org.apache.kafka.clients.producer.KafkaProducer;
+import org.apache.kafka.clients.producer.ProducerRecord;
+import org.apache.kafka.common.TopicPartition;
+import org.junit.AfterClass;
+import org.junit.Before;
+import org.junit.BeforeClass;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.TestName;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Properties;
+import java.util.Set;
+
+import static com.netease.arctic.flink.shuffle.RowKindUtil.transformFromFlinkRowKind;
+import static com.netease.arctic.flink.table.descriptors.ArcticValidator.ARCTIC_LOG_CONSISTENCY_GUARANTEE_ENABLE;
+import static com.netease.arctic.flink.util.kafka.KafkaContainerTest.KAFKA_CONTAINER;
+import static com.netease.arctic.flink.util.kafka.KafkaContainerTest.getPropertiesByTopic;
+import static com.netease.arctic.flink.util.kafka.KafkaContainerTest.readRecordsBytes;
+import static com.netease.arctic.flink.write.hidden.TestBaseLog.createLogDataDeserialization;
+import static com.netease.arctic.flink.write.hidden.TestBaseLog.userSchema;
+import static com.netease.arctic.flink.write.hidden.TestHiddenLogOperators.createRowData;
+import static org.junit.Assert.assertEquals;
+
+public class TestKafkaSourceReader {
+  private static final Logger LOG = LoggerFactory.getLogger(TestKafkaSourceReader.class);
+  private static String topic;
+  private static final int NUM_SPLITS = 1;
+  private static final int NUM_RECORDS_PER_SPLIT = 10;
+  private static final int TOTAL_NUM_RECORDS = NUM_RECORDS_PER_SPLIT * NUM_SPLITS;
+
+  @Rule
+  public TestName testName = new TestName();
+
+  private static final byte[] JOB_ID = IdGenerator.generateUpstreamId();
+
+  @BeforeClass
+  public static void prepare() throws Exception {
+    KAFKA_CONTAINER.start();
+  }
+
+  @AfterClass
+  public static void shutdown() throws Exception {
+    KAFKA_CONTAINER.close();
+  }
+
+  @Before
+  public void initData() throws Exception {
+    topic = TestUtil.getUtMethodName(testName);
```

Comment:
How many partitions does this topic have?

Reply:
The default number of partitions is 1, which I have changed to be explicitly specified in the unit test.